### ISSUE TITLE
[SPIR-V] Unstripped OpCapability ImageMSArray

### ISSUE THREAD

**Issue by tteguh** [microsoft/DirectXShaderCompiler#4795]
I have the following HLSL shader with two possible entry point:  One uses a `Texture2DMSArray`, the other uses a `Texture2DMS`

```
Texture2DMS<float4, 2> mainTexture;
Texture2DMSArray<float4, 2> mainTextureArray;

float4 psTex2d() : SV_Target
{
	float4 c = mainTexture.Load(int2(0,0),0);
	return c;
}

float4 psTex2dArray() : SV_Target
{
	float4 c = mainTextureArray.Load(int3(0,0,0),0);
	return c;
}
```

As far as I'm aware, `OpCapability ImageMSArray` is only needed when `Texture2DMSArray` is used.  DXC will strip out the unnecessary texture resource when compiled with `-O3` depending on the passed in entry point.  However, the capability is still added to the resulting SPIRV, regardless of which entry point was used:

`dxc.exe -spirv -O3 -T ps_6_0 -E psTex2d`:
```
; SPIR-V
; Version: 1.0
; Generator: Google spiregg; 0
; Bound: 18
; Schema: 0
               OpCapability Shader
               OpCapability ImageMSArray
               OpMemoryModel Logical GLSL450
               OpEntryPoint Fragment %psTex2d "psTex2d" %out_var_SV_Target
               OpExecutionMode %psTex2d OriginUpperLeft
               OpSource HLSL 600
               OpName %type_2d_image "type.2d.image"
               OpName %mainTexture "mainTexture"
               OpName %out_var_SV_Target "out.var.SV_Target"
               OpName %psTex2d "psTex2d"
               OpDecorate %out_var_SV_Target Location 0
               OpDecorate %mainTexture DescriptorSet 0
               OpDecorate %mainTexture Binding 0
        %int = OpTypeInt 32 1
      %int_0 = OpConstant %int 0
      %v2int = OpTypeVector %int 2
          %8 = OpConstantComposite %v2int %int_0 %int_0
      %float = OpTypeFloat 32
%type_2d_image = OpTypeImage %float 2D 2 0 1 1 Unknown
%_ptr_UniformConstant_type_2d_image = OpTypePointer UniformConstant %type_2d_image
    %v4float = OpTypeVector %float 4
%_ptr_Output_v4float = OpTypePointer Output %v4float
       %void = OpTypeVoid
         %14 = OpTypeFunction %void
%mainTexture = OpVariable %_ptr_UniformConstant_type_2d_image UniformConstant
%out_var_SV_Target = OpVariable %_ptr_Output_v4float Output
    %psTex2d = OpFunction %void None %14
         %15 = OpLabel
         %16 = OpLoad %type_2d_image %mainTexture
         %17 = OpImageFetch %v4float %16 %8 Sample %int_0
               OpStore %out_var_SV_Target %17
               OpReturn
               OpFunctionEnd
```

`dxc.exe -spirv -O3 -T ps_6_0 -E psTex2dArray`:
```
; SPIR-V
; Version: 1.0
; Generator: Google spiregg; 0
; Bound: 18
; Schema: 0
               OpCapability Shader
               OpCapability ImageMSArray
               OpMemoryModel Logical GLSL450
               OpEntryPoint Fragment %psTex2dArray "psTex2dArray" %out_var_SV_Target
               OpExecutionMode %psTex2dArray OriginUpperLeft
               OpSource HLSL 600
               OpName %type_2d_image_array "type.2d.image.array"
               OpName %mainTextureArray "mainTextureArray"
               OpName %out_var_SV_Target "out.var.SV_Target"
               OpName %psTex2dArray "psTex2dArray"
               OpDecorate %out_var_SV_Target Location 0
               OpDecorate %mainTextureArray DescriptorSet 0
               OpDecorate %mainTextureArray Binding 1
        %int = OpTypeInt 32 1
      %int_0 = OpConstant %int 0
      %v3int = OpTypeVector %int 3
          %8 = OpConstantComposite %v3int %int_0 %int_0 %int_0
      %float = OpTypeFloat 32
%type_2d_image_array = OpTypeImage %float 2D 2 1 1 1 Unknown
%_ptr_UniformConstant_type_2d_image_array = OpTypePointer UniformConstant %type_2d_image_array
    %v4float = OpTypeVector %float 4
%_ptr_Output_v4float = OpTypePointer Output %v4float
       %void = OpTypeVoid
         %14 = OpTypeFunction %void
%mainTextureArray = OpVariable %_ptr_UniformConstant_type_2d_image_array UniformConstant
%out_var_SV_Target = OpVariable %_ptr_Output_v4float Output
%psTex2dArray = OpFunction %void None %14
         %15 = OpLabel
         %16 = OpLoad %type_2d_image_array %mainTextureArray
         %17 = OpImageFetch %v4float %16 %8 Sample %int_0
               OpStore %out_var_SV_Target %17
               OpReturn
               OpFunctionEnd

```

I would've expected DXC to strip out `OpCapability ImageMSArray` when it stripped out `mainTextureArray` from the shader compiled with `psTex2d` as the entry point but that doesn't seem to be the case.  

This is problematic for devices that only use the `psTex2d` path and does not support `OpCapability ImageMSArray` as `shaderStorageImageMultisample` is not available on the device.  Vulkan validation will check this at runtime and trigger a validation error message.

I can reproduce this using the latest build at commit: `46708130968ed1e1e0beacd7f1187d63eeb48915`

**Comment by vettoreldaniele** [comment]
We visit instructions to add capabilities before we do any optimizations, hence this problem. We need to look more into it and understand edge cases, but maybe it's worth re-doing a pass on needed capabilities after optimizations have been run and then remove those that are not needed anymore.

**Comment by s-perron** [comment]
This is not a simple change. As Daniele mentioned, capabilities are added before optimizations. In particular, capabilities are added while the code is still in DXC's internal format. It is then converted to the official spir-v binary format to be passed to the optimizer. There is currently no way to read a spir-v binary to put it into DXC's internal format, so we are not likely to reuse DXC's code that will tell us which capabilities are required.

SPIR-V Tools has some of the logic to know which capabilities are required for each instruction in the validator, but that code is not easily shared with the optimizer. I'll look into creating a spirv-opt pass that will remove unneeded capabilities and extensions, but, to do this right, I want to find a way to refactor the code to share the logic with the validator.

**Comment by s-perron** [comment]
I looked at the validator code yesterday, and it will be very hard to refactor it. The process used by the validator is very different than what we will want in spirv-opt.

My current idea, which I will not get to right away, is to have a pass in spirv-opt that will remove the capabilities that are **known** to not be needed.

The pass must have a list of known capabilities.

**Comment by s-perron** [comment]
We are trying to implement a pass to remove unneeded capabilities. The first phase is https://github.com/KhronosGroup/SPIRV-Tools/pull/5278.

**Comment by natevm** [comment]
> We are trying to implement a pass to remove unneeded capabilities. The first phase is [KhronosGroup/SPIRV-Tools#5278](https://github.com/KhronosGroup/SPIRV-Tools/pull/5278).

Now that this issue is merged into SPIRV tools, what is the next phase towards removing unnecessary capability requests?

**Comment by s-perron** [comment]
We need to call the pass from DXC, and improve the pass to handle more capabilities. If you give us a list of capabilities which you need removed, we could get to those earlier.

**Comment by natevm** [comment]
> We need to call the pass from DXC, and improve the pass to handle more capabilities. If you give us a list of capabilities which you need removed, we could get to those earlier.

The main two capabilities that I’m blocked on until they get stripped are SPV_KHR_ray_query and SPV_KHR_ray_tracing. If we could get those fixed, we could start doing light ray tracing work on NV pascal, some heavier simulation stuff on these A100 systems at ANL, and then open the door to some HLSL -> SPV -> MoltenVK/Metal RT support (currently only ray queries there are intended to work)

Down the road I might also hit some snags with SPV_KHR_cooperative_matrix, but ray tracing versus ray query are the main two that are causing me grief.

**Comment by tteguh** [comment]
Thanks for looking into this @s-perron and @Keenuts.  Commenting to make sure that the original issue here doesn't get stale. We're still interested in getting ImageMSArray capability removed when it's unnecessary, so if the team could get to those earlier that would be very helpful for us

**Comment by Keenuts** [comment]
Hello! Thanks, I'll try to make those capabilities a priority.
Right now, I'm working on integrating the pass into DXC (few bugs, cases to add), and once integrated adding new capabilities shouldn't take too much time.

**Comment by natevm** [comment]
@Keenuts With PR #5518 merged into master, does this mean we can test and see if the unused capabilities mentioned here get removed with the latest DXC version? Not sure how far along that is, but maybe we can close this issue if everything's hooked up now.

**Comment by Keenuts** [comment]
Hello!
I have some free time to focus on that now.

> @Keenuts With PR https://github.com/microsoft/DirectXShaderCompiler/pull/5518 merged into master, does this mean we can test and see if the unused capabilities mentioned here get removed with the latest DXC version?

Not yet!
The pass did only brought a few capabilities (https://github.com/KhronosGroup/SPIRV-Tools/blob/main/source/opt/trim_capabilities_pass.h#L75)
But running this pass do change slightly the behavior of DXC (capability names with the same ID can change, and some module ordering can change, while remaining semantically the same).
For this reason, I wanted to let the basic pass (with very few capabilities supported) soak in a bit.

In the meantime, I'm new capabilities to SPIR-V tools. `ImageMSArray` was the first on my list, and not I'll look into ray query/tracing capabilities.

**Comment by Keenuts** [comment]
The capability `ImageMSArray` is now correctly trimmed at head, so this should be fixed 😊
